### PR TITLE
Implement Cloning multiple copies

### LINKED ISSUE #5: As a computing student who has many software subscriptions, I can use the recurring expenses tracker
... so that I can manage my subscription payments.

### LINKED ISSUE #18: As a NUS international student, I need to be able to keep track of recurring utility expenses such as phone bills
... so that I can plan a suitable budget

### LINKED ISSUE #96: As a student who has repeated transactions that are completely or almost identical, I want to make copies of such transactions
... so that it is more convenient to add them instead of having to type them out fully each time.

### LINKED ISSUE #183: Enhance Clone command for creating multiple clones with certain frequency
Extend `clone` command to allow creating multiple clones over multiple days, weeks, months or years.

Increment of #58

### PR DESCRIPTION
**Pull request by lye-jw**
Add ability to Clone multiple copies of a **`Transaction`** with command **clone i/INDEX o/OCCURRENCE**, where **OCCURRENCE** is in the format **FREQUENCY:NUMBER_OF_OCCURRENCES**.

- **FREQUENCY** refers to how much time is between the clones: `daily`, `weekly`, `monthly`, `yearly`.
- **NUMBER_OF_OCCURRENCES** refers to how many times the Transaction is cloned.

Tests for new **`Occurrence`** class and Clone functionality included.
(Updates to documentation to be done later, probably in another PR)

*Ignore the following tags:*
Resolves #183 Resolves #5 Resolves #18 Resolves #96

### CODE REVIEW

File: src/main/java/thrift/logic/commands/exceptions/CommandException.java [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package thrift.logic.commands.exceptions;
 
+import thrift.logic.commands.Command;
```
Comment (limerencee) [inline src/main/java/thrift/logic/commands/exceptions/CommandException.java:3]: Is this import used? :open_mouth:

**Review by limerencee — CHANGES_REQUESTED** [review]
LGTM after minor clarification!

File: src/main/java/thrift/logic/commands/exceptions/CommandException.java (hunk 1 above)
Comment (lye-jw) [reply src/main/java/thrift/logic/commands/exceptions/CommandException.java:3]: Removed in latest update!

**Comment by lightz96** [comment]
Should we limit the number of occurrences to a smaller number?

**Comment by lye-jw** [comment]
> Should we limit the number of frequencies to a smaller number?

Good idea, what would be a good number? 7 so users can populate a week with `daily`, or 12 so users can populate a year with `monthly`? Or even less?

**Comment by limerencee** [comment]
> > Should we limit the number of frequencies to a smaller number?
> 
> Good idea, what would be a good number? 7 so users can populate a week with `daily`, or 12 so users can populate a year with `monthly`? Or even less?

12 sounds like a reasonable upper bound. :thinking:

**Comment by lightz96** [comment]
> > Should we limit the number of frequencies to a smaller number?
> 
> Good idea, what would be a good number? 7 so users can populate a week with `daily`, or 12 so users can populate a year with `monthly`? Or even less?

Probably use different limits for each of the frequencies? E.g. 7 for `daily`, 12 for `monthly`, probably 5 for `yearly`? We need some reasons in case the reviewer starts to ask for the reason.

**Comment by lye-jw** [comment]
> > > Should we limit the number of frequencies to a smaller number?
> > 
> > 
> > Good idea, what would be a good number? 7 so users can populate a week with `daily`, or 12 so users can populate a year with `monthly`? Or even less?
> 
> Sounds good to me. 7 for `daily`, 12 for `monthly`, probably 5 for `yearly`? We need some reasons in case the reviewer starts to ask for the reason.

`yearly` maxed at 5 is good. But I think everything else should allow up to 12 so user can have freedom in adding more days and weeks.

**Comment by lightz96** [comment]
> > > > Should we limit the number of frequencies to a smaller number?
> > > 
> > > 
> > > Good idea, what would be a good number? 7 so users can populate a week with `daily`, or 12 so users can populate a year with `monthly`? Or even less?
> > 
> > 
> > Sounds good to me. 7 for `daily`, 12 for `monthly`, probably 5 for `yearly`? We need some reasons in case the reviewer starts to ask for the reason.
> 
> `yearly` maxed at 5 is good. But I think everything else should allow up to 12 so user can have freedom in adding more days and weeks.

Ok, sounds good to me.

**Comment by lightz96** [comment]
If I execute **clone i/1 o/monthly:5** on 31 October, it adds to 29 February because February does not have 31 days (this also applies to months without 31 days), should we allow this to happen? 🤔

**Comment by lye-jw** [comment]
> If I execute **clone i/1 o/monthly:5** on 31 October, it adds to 29 February because February does not have 31 days (this also applies to months without 31 days), should we allow this to happen? 🤔

Java's `Calendar#add` correctly handles day overflowing for months. 2020 is a leap year so February 2020 has 29 days. From [Calendar documentation](https://docs.oracle.com/javase/6/docs/api/java/util/Calendar.html):

```
Add rule 1. The value of field f after the call minus the value of field f before the call is delta, modulo any overflow that has occurred in field f. Overflow occurs when a field value exceeds its range and, as a result, the next larger field is incremented or decremented and the field value is adjusted back into its range.

Add rule 2. If a smaller field is expected to be invariant, but it is impossible for it to be equal to its prior value because of changes in its minimum or maximum after field f is changed or other constraints, such as time zone offset changes, then its value is adjusted to be as close as possible to its expected value. A smaller field represents a smaller unit of time. HOUR is a smaller field than DAY_OF_MONTH. No adjustment is made to smaller fields that are not expected to be invariant. The calendar system determines what fields are expected to be invariant.
```

**Comment by lye-jw** [comment]
> If I execute **clone i/1 o/monthly:5** on 31 October, it adds to 29 February because February does not have 31 days (this also applies to months without 31 days), should we allow this to happen? 🤔

That's an interesting point. We have to think about why the transaction is to be repeated and what happens to it for months that don't have those days.

My opinion is to let them add to the last possible day, since most examples I can think of are transactions that occur at the end of the month regardless of specific day, such as salary payday or monthly bill payment.

What does everyone think? @lightz96 @limerencee @krusagiz @kensonoenqh

**Comment by lightz96** [comment]
> > If I execute **clone i/1 o/monthly:5** on 31 October, it adds to 29 February because February does not have 31 days (this also applies to months without 31 days), should we allow this to happen? 🤔
> 
> Java's `Calendar#add` correctly handles day overflowing for months. 2020 is a leap year so February 2020 has 29 days. From [Calendar documentation](https://docs.oracle.com/javase/6/docs/api/java/util/Calendar.html):
> 
> ```
> Add rule 1. The value of field f after the call minus the value of field f before the call is delta, modulo any overflow that has occurred in field f. Overflow occurs when a field value exceeds its range and, as a result, the next larger field is incremented or decremented and the field value is adjusted back into its range.
> 
> Add rule 2. If a smaller field is expected to be invariant, but it is impossible for it to be equal to its prior value because of changes in its minimum or maximum after field f is changed or other constraints, such as time zone offset changes, then its value is adjusted to be as close as possible to its expected value. A smaller field represents a smaller unit of time. HOUR is a smaller field than DAY_OF_MONTH. No adjustment is made to smaller fields that are not expected to be invariant. The calendar system determines what fields are expected to be invariant.
> ```

Should we add the transaction with the date 29 February or skip it? (for this scenario)

**Comment by lye-jw** [comment]
> > > If I execute **clone i/1 o/monthly:5** on 31 October, it adds to 29 February because February does not have 31 days (this also applies to months without 31 days), should we allow this to happen? 🤔
> > 
> > 
> > Java's `Calendar#add` correctly handles day overflowing for months. 2020 is a leap year so February 2020 has 29 days. From [Calendar documentation](https://docs.oracle.com/javase/6/docs/api/java/util/Calendar.html):
> > ```
> > Add rule 1. The value of field f after the call minus the value of field f before the call is delta, modulo any overflow that has occurred in field f. Overflow occurs when a field value exceeds its range and, as a result, the next larger field is incremented or decremented and the field value is adjusted back into its range.
> > 
> > Add rule 2. If a smaller field is expected to be invariant, but it is impossible for it to be equal to its prior value because of changes in its minimum or maximum after field f is changed or other constraints, such as time zone offset changes, then its value is adjusted to be as close as possible to its expected value. A smaller field represents a smaller unit of time. HOUR is a smaller field than DAY_OF_MONTH. No adjustment is made to smaller fields that are not expected to be invariant. The calendar system determines what fields are expected to be invariant.
> > ```
> 
> Should we add the transaction with the date 29 February or skip it?

Sorry I misunderstood your comment at first, this reply is outdated. See my latest reply:

> That's an interesting point. We have to think about why the transaction is to be repeated and what happens to it for months that don't have those days.
> 
> My opinion is to let them add to the last possible day, since most examples I can think of are transactions that occur at the end of the month regardless of specific day, such as salary payday or monthly bill payment.
> 
> What does everyone think? @lightz96 @limerencee @krusagiz @kensonoenqh

**Comment by lightz96** [comment]
> > > > If I execute **clone i/1 o/monthly:5** on 31 October, it adds to 29 February because February does not have 31 days (this also applies to months without 31 days), should we allow this to happen? 🤔
> > > 
> > > 
> > > Java's `Calendar#add` correctly handles day overflowing for months. 2020 is a leap year so February 2020 has 29 days. From [Calendar documentation](https://docs.oracle.com/javase/6/docs/api/java/util/Calendar.html):
> > > ```
> > > Add rule 1. The value of field f after the call minus the value of field f before the call is delta, modulo any overflow that has occurred in field f. Overflow occurs when a field value exceeds its range and, as a result, the next larger field is incremented or decremented and the field value is adjusted back into its range.
> > > 
> > > Add rule 2. If a smaller field is expected to be invariant, but it is impossible for it to be equal to its prior value because of changes in its minimum or maximum after field f is changed or other constraints, such as time zone offset changes, then its value is adjusted to be as close as possible to its expected value. A smaller field represents a smaller unit of time. HOUR is a smaller field than DAY_OF_MONTH. No adjustment is made to smaller fields that are not expected to be invariant. The calendar system determines what fields are expected to be invariant.
> > > ```
> > 
> > 
> > Should we add the transaction with the date 29 February or skip it?
> 
> Sorry I misunderstood your comment at first, see my latest previous reply:
> 
> > That's an interesting point. We have to think about why the transaction is to be repeated and what happens to it for months that don't have those days.
> > My opinion is to let them add to the last possible day, since most examples I can think of are transactions that occur at the end of the month regardless of specific day, such as salary payday or monthly bill payment.
> > What does everyone think? @lightz96 @limerencee @krusagiz @kensonoenqh

Good examples that you have come up with. Therefore, I agreed with the fact it should be added to the end of the month regardless of the specific day.

**Review by limerencee — APPROVED** [review]

**Comment by krusagiz** [comment]
> > If I execute **clone i/1 o/monthly:5** on 31 October, it adds to 29 February because February does not have 31 days (this also applies to months without 31 days), should we allow this to happen? 🤔
> 
> That's an interesting point. We have to think about why the transaction is to be repeated and what happens to it for months that don't have those days.
> 
> My opinion is to let them add to the last possible day, since most examples I can think of are transactions that occur at the end of the month regardless of specific day, such as salary payday or monthly bill payment.
> 
> What does everyone think? @lightz96 @limerencee @krusagiz @kensonoenqh

The PR has been closed but I agree with @lye-jw that it should add to the last possible day. This way, we can handle a recurrence in every last day of the month, just by putting the biggest possible day or an arbitrarily large number.

**Pull request merged**